src/Configuration/Config.AzureKeyVault/src/AzureKeyVaultConfigurationProvider.cs

```diff
@@ -31,67 +30,99 @@ internal class AzureKeyVaultConfigurationProvider : ConfigurationProvider
-
-            if (reloadOnChange && reloadPollDelay <= 0)
+            if (_reloadPollDelay != null && _reloadPollDelay.Value == TimeSpan.Zero)
             {
-                throw new ArgumentException("{0} must be greater than 0", nameof(reloadPollDelay));
+                throw new ArgumentException(nameof(reloadPollDelay));
             }
 
             _pollingTask = null;
             _cancellationToken = new CancellationTokenSource();
-            _reloadOnChange = reloadOnChange;
             _reloadPollDelay = reloadPollDelay;
             _loadedSecrets = new Dictionary<string, SecretAttributes>();
         }
 
-        ~AzureKeyVaultConfigurationProvider()
-        {
-            _cancellationToken.Cancel();
-        }
-
         public override void Load() => LoadAsync().ConfigureAwait(false).GetAwaiter().GetResult();
 
-        private async Task PollForSecretChangesAsync(int reloadPollDelay, CancellationToken ct)
+        private async Task PollForSecretChangesAsync(CancellationToken ct)
         {
             while (!ct.IsCancellationRequested)
             {
-                await Task.Delay(reloadPollDelay);
-                var secretsHaveChanged = await ShouldReloadAsync();
+                await Task.Delay(_reloadPollDelay.Value.Milliseconds);
+                Load();
+            }
+        }
 
-                // if the secret list has changed, reload.
-                if(secretsHaveChanged)
-                {
-                    Load();
-                }
+        private bool ShouldReloadSecret(SecretItem secretItem)
+        {
+            bool result = false;
+
+            if (_manager.Load(secretItem) && secretItem.Attributes?.Enabled == true)
+            {
+                string key = secretItem.Identifier.Name;
+                var secretUpdateTime = secretItem.Attributes.Updated;
+                var cachedUpdateTime = _loadedSecrets[key]?.Updated;
+                bool isKeyLoaded = _loadedSecrets.ContainsKey(key);
+                bool isUpdateTimeknown = (secretUpdateTime != null) && (cachedUpdateTime != null);
+                bool isKeyUnchanged = isKeyLoaded &&
+                                        isUpdateTimeknown &&
+                                        (cachedUpdateTime?.Ticks == secretUpdateTime?.Ticks);
+                result = !isKeyUnchanged;
             }
+            return result;
         }
 
         private async Task LoadAsync()
         {
             var data = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
-
-            _loadedSecrets.Clear();
-
             var secrets = await _client.GetSecretsAsync(_vault).ConfigureAwait(false);
+            var secretList = new List<SecretItem>(secrets.Count());
             var tasks = new List<Task<SecretBundle>>(secrets.Count());
+            bool isReloadNeeded = false;
+
+            // if a secret has been removed/added, mark for reloading
+            if (secrets.Count() != _loadedSecrets.Count())
```

Comment:
This has to be done after the do/while loop since `secrets` is paged and won't have the full list. Use `secretList` instead.